File: packages/compiler/tw/lexer/strings.ts
/**
 * String escape handling -- full Unicode, template literals, multi-line,
 * all JavaScript escape sequences including \u{XXXX}, \xXX, \0, etc.
 */

export interface StringReadResult {
  value: string;
  raw: string;
  end: number;
  hasNewline: boolean;
  isTemplate: boolean;
  expressions: TemplateExpression[];
  error?: string;
}

export interface TemplateExpression {
  start: number;
  end: number;
  expression: string;
}

const ESCAPE_MAP: Record<string, string> = {
  "n": "\n",
  "r": "\r",
  "t": "\t",
  "b": "\b",
  "f": "\f",
  "v": "\v",
  "0": "\0",
  "\\": "\\",
  "'": "'",
  '"': '"',
  "`": "`",
  "$": "$",
  "/": "/",
};

// Unicode line terminators
const LINE_TERMINATORS = new Set(["\n", "\r", "\u2028", "\u2029"]);

export function readString(
  source: string,
  start: number,
  quote: string = '"',
): StringReadResult {
  if (source[start] !== quote) {
    return { value: "", raw: "", end: start, hasNewline: false, isTemplate: false, expressions: [], error: `Expected ${quote}` };
  }

  let i = start + 1;
  let value = "";
  let hasNewline = false;
  const expressions: TemplateExpression[] = [];
  const isTemplate = quote === "`";

  while (i < source.length) {
    const ch = source[i];

    // Closing quote
    if (ch === quote) {
      if (!isTemplate || source[i - 1] !== "\\") {
        return {
          value,
          raw: source.slice(start, i + 1),
          end: i + 1,
          hasNewline,
          isTemplate,
          expressions,
        };
      }
    }

    // Template literal expression
    if (isTemplate && ch === "$" && source[i + 1] === "{") {
      const exprStart = i + 2;
      const exprResult = readTemplateExpression(source, exprStart);
      if (exprResult.end === -1) {
        return {
          value,
          raw: source.slice(start, i),
          end: source.length,
          hasNewline,
          isTemplate,
          expressions,
          error: "Unclosed template expression",
        };
      }
      expressions.push({
        start: exprStart,
        end: exprResult.end,
        expression: exprResult.text,
      });
      value += `\x00${expressions.length - 1}\x00`;
      i = exprResult.end + 1;
      continue;
    }

    // Escape sequences
    if (ch === "\\") {
      const escapeResult = readEscape(source, i);
      if (escapeResult.error) {
        return {
          value,
          raw: source.slice(start, i),
          end: i,
          hasNewline,
          isTemplate,
          expressions,
          error: escapeResult.error,
        };
      }
      value += escapeResult.value;
      if (LINE_TERMINATORS.has(escapeResult.value)) {
        hasNewline = true;
      }
      i = escapeResult.end;
      continue;
    }

    // Line terminators (illegal in non-template strings)
    if (LINE_TERMINATORS.has(ch)) {
      if (!isTemplate) {
        return {
          value,
          raw: source.slice(start, i),
          end: i,
          hasNewline: true,
          isTemplate,
          expressions,
          error: "Unterminated string literal",
        };
      }
      hasNewline = true;
    }

    value += ch;
    i++;
  }

  return {
    value,
    raw: source.slice(start, i),
    end: i,
    hasNewline,
    isTemplate,
    expressions,
    error: "Unterminated string literal",
  };
}

function readTemplateExpression(source: string, start: number): { text: string; end: number } {
  let i = start;
  let depth = 1;
  let text = "";

  while (i < source.length) {
    const ch = source[i];

    if (ch === "{") depth++;
    else if (ch === "}") {
      depth--;
      if (depth === 0) {
        return { text, end: i };
      }
    }

    // Skip nested strings
    if (ch === '"' || ch === "'" || ch === "`") {
      const strResult = readString(source, i, ch);
      text += strResult.raw;
      i = strResult.end;
      continue;
    }

    text += ch;
    i++;
  }

  return { text, end: -1 };
}

interface EscapeResult {
  value: string;
  end: number;
  error?: string;
}

function readEscape(source: string, start: number): EscapeResult {
  const ch = source[start + 1];

  if (ch === undefined) {
    return { value: "\\", end: start + 1, error: "Unterminated escape sequence" };
  }

  // Simple escapes
  if (ESCAPE_MAP[ch]) {
    return { value: ESCAPE_MAP[ch], end: start + 2 };
  }

  // Unicode escape \uXXXX
  if (ch === "u") {
    // \u{XXXXX} -- Unicode code point escape
    if (source[start + 2] === "{") {
      const end = source.indexOf("}", start + 3);
      if (end === -1) {
        return { value: "", end: start + 2, error: "Unterminated Unicode code point escape" };
      }
      const hex = source.slice(start + 3, end);
      const codePoint = parseInt(hex, 16);
      if (isNaN(codePoint) || codePoint > 0x10FFFF) {
        return { value: "", end, error: "Invalid Unicode code point" };
      }
      return { value: String.fromCodePoint(codePoint), end: end + 1 };
    }

    // \uXXXX -- Fixed-length Unicode escape
    const hex = source.slice(start + 2, start + 6);
    if (!/^[0-9a-fA-F]{4}$/.test(hex)) {
      return { value: "", end: start + 2, error: "Invalid Unicode escape sequence" };
    }
    const codePoint = parseInt(hex, 16);
    return { value: String.fromCharCode(codePoint), end: start + 6 };
  }

  // Hex escape \xXX
  if (ch === "x") {
    const hex = source.slice(start + 2, start + 4);
    if (!/^[0-9a-fA-F]{2}$/.test(hex)) {
      return { value: "", end: start + 2, error: "Invalid hex escape sequence" };
    }
    const codePoint = parseInt(hex, 16);
    return { value: String.fromCharCode(codePoint), end: start + 4 };
  }

  // Octal escape (deprecated but supported)
  if (/^[0-7]$/.test(ch)) {
    let octal = ch;
    let i = start + 2;
    while (i < source.length && /^[0-7]$/.test(source[i]) && octal.length < 3) {
      octal += source[i];
      i++;
    }
    const codePoint = parseInt(octal, 8);
    if (codePoint > 255) {
      return { value: "", end: i, error: "Octal escape out of range" };
    }
    return { value: String.fromCharCode(codePoint), end: i };
  }

  // Null escape
  if (ch === "0" && !/^[0-9]/.test(source[start + 2] ?? "")) {
    return { value: "\0", end: start + 2 };
  }

  // Line continuation (backslash followed by line terminator)
  if (LINE_TERMINATORS.has(ch)) {
    return { value: "", end: start + 2 };
  }

  // Unknown escape -- keep the character as-is
  return { value: ch, end: start + 2 };
}

// --- String Validation ------------------------------------------------------

export function validateString(str: string): string[] {
  const errors: string[] = [];

  // Check for invalid control characters
  for (let i = 0; i < str.length; i++) {
    const code = str.charCodeAt(i);
    if (code < 0x20 && code !== 0x09 && code !== 0x0A && code !== 0x0D) {
      errors.push(`Invalid control character at position ${i}: \\x${code.toString(16).padStart(2, "0")}`);
    }
  }

  return errors;
}

// --- String Utilities --------------------------------------------------------

export function unescapeString(raw: string): string {
  if (!raw.startsWith('"') && !raw.startsWith("'") && !raw.startsWith("`")) {
    return raw;
  }
  const quote = raw[0];
  const inner = raw.slice(1, raw.endsWith(quote) ? -1 : undefined);
  let result = "";
  let i = 0;

  while (i < inner.length) {
    if (inner[i] === "\\" && i + 1 < inner.length) {
      const escapeResult = readEscape(inner, i);
      result += escapeResult.value;
      i = escapeResult.end;
    } else {
      result += inner[i];
      i++;
    }
  }

  return result;
}

export function escapeString(str: string, quote: string = '"'): string {
  let result = quote;

  for (let i = 0; i < str.length; i++) {
    const ch = str[i];

    switch (ch) {
      case "\\": result += "\\\\"; break;
      case quote: result += `\\${ch}`; break;
      case "\n": result += "\\n"; break;
      case "\r": result += "\\r"; break;
      case "\t": result += "\\t"; break;
      case "\b": result += "\\b"; break;
      case "\f": result += "\\f"; break;
      case "\v": result += "\\v"; break;
      case "\0": result += "\\0"; break;
      default:
        const code = ch.charCodeAt(0);
        if (code < 0x20) {
          result += `\\x${code.toString(16).padStart(2, "0")}`;
        } else if (code > 0xFFFF) {
          // Supplementary plane -- use \u{XXXXX}
          result += `\\u{${code.toString(16)}}`;
        } else if (code > 0x7E) {
          result += `\\u${code.toString(16).padStart(4, "0")}`;
        } else {
          result += ch;
        }
    }
  }

  result += quote;
  return result;
}

export function isStringDelimiter(ch: string): boolean {
  return ch === '"' || ch === "'" || ch === "`";
}

export function getStringDelimiter(str: string): string | null {
  if (str.length === 0) return null;
  return isStringDelimiter(str[0]) ? str[0] : null;
}

// --- Regex Reading -----------------------------------------------------------

export interface RegexReadResult {
  pattern: string;
  flags: string;
  raw: string;
  end: number;
  error?: string;
}

export function readRegex(source: string, start: number): RegexReadResult {
  if (source[start] !== "/") {
    return { pattern: "", flags: "", raw: "", end: start, error: "Expected /" };
  }

  let i = start + 1;
  let pattern = "";
  let inClass = false;

  while (i < source.length) {
    const ch = source[i];

    if (ch === "\\") {
      pattern += ch;
      if (i + 1 < source.length) {
        pattern += source[i + 1];
        i += 2;
        continue;
      }
    }

    if (ch === "[") inClass = true;
    if (ch === "]") inClass = false;

    if (ch === "/" && !inClass) {
      // End of pattern -- read flags
      i++;
      let flags = "";
      while (i < source.length && /[gimsuyd]/.test(source[i])) {
        flags += source[i];
        i++;
      }
      return { pattern, flags, raw: source.slice(start, i), end: i };
    }

    if (ch === "\n" || ch === "\r") {
      return { pattern, flags: "", raw: source.slice(start, i), end: i, error: "Unterminated regex literal" };
    }

    pattern += ch;
    i++;
  }

  return { pattern, flags: "", raw: source.slice(start, i), end: i, error: "Unterminated regex literal" };
}

// --- Number Reading ---------------------------------------------------------

export interface NumberReadResult {
  value: string;
  raw: string;
  end: number;
  isFloat: boolean;
  isBigInt: boolean;
  radix: number;
  error?: string;
}

export function readNumber(source: string, start: number): NumberReadResult {
  let i = start;
  let value = "";
  let isFloat = false;
  let isBigInt = false;
  let radix = 10;

  // Check for hex, octal, binary
  if (source[i] === "0" && i + 1 < source.length) {
    const next = source[i + 1].toLowerCase();
    if (next === "x") {
      radix = 16;
      i += 2;
      while (i < source.length && /[0-9a-fA-F_]/.test(source[i])) {
        value += source[i];
        i++;
      }
      return { value, raw: source.slice(start, i), end: i, isFloat: false, isBigInt, radix };
    }
    if (next === "o") {
      radix = 8;
      i += 2;
      while (i < source.length && /[0-7_]/.test(source[i])) {
        value += source[i];
        i++;
      }
      return { value, raw: source.slice(start, i), end: i, isFloat: false, isBigInt, radix };
    }
    if (next === "b") {
      radix = 2;
      i += 2;
      while (i < source.length && /[01_]/.test(source[i])) {
        value += source[i];
        i++;
      }
      return { value, raw: source.slice(start, i), end: i, isFloat: false, isBigInt, radix };
    }
  }

  // Decimal number
  while (i < source.length && /[0-9_]/.test(source[i])) {
    value += source[i];
    i++;
  }

  // Fractional part
  if (source[i] === "." && /[0-9]/.test(source[i + 1] ?? "")) {
    isFloat = true;
    value += ".";
    i++;
    while (i < source.length && /[0-9_]/.test(source[i])) {
      value += source[i];
      i++;
    }
  }

  // Exponent
  if (source[i] === "e" || source[i] === "E") {
    isFloat = true;
    value += source[i];
    i++;
    if (source[i] === "+" || source[i] === "-") {
      value += source[i];
      i++;
    }
    while (i < source.length && /[0-9_]/.test(source[i])) {
      value += source[i];
      i++;
    }
  }

  // BigInt suffix
  if (source[i] === "n") {
    isBigInt = true;
    i++;
  }

  return { value, raw: source.slice(start, i), end: i, isFloat, isBigInt, radix };
}

File: rust/crates/tw-lexer/src/strings.rs
/// String, number, and regex reading — ported from the TS lexer/strings.ts.
/// Handles all JS escape sequences, template literals, hex/octal/binary numbers.

#[derive(Debug, Clone)]
pub struct StringReadResult {
    pub value: String,
    pub raw: String,
    pub end: usize,
    pub has_newline: bool,
    pub is_template: bool,
    pub expressions: Vec<TemplateExpression>,
    pub error: Option<String>,
}

#[derive(Debug, Clone)]
pub struct TemplateExpression {
    pub start: usize,
    pub end: usize,
    pub expression: String,
}

#[derive(Debug, Clone)]
pub struct NumberReadResult {
    pub value: String,
    pub raw: String,
    pub end: usize,
    pub is_float: bool,
    pub is_big_int: bool,
    pub radix: u32,
    pub error: Option<String>,
}

#[derive(Debug, Clone)]
pub struct RegexReadResult {
    pub pattern: String,
    pub flags: String,
    pub raw: String,
    pub end: usize,
    pub error: Option<String>,
}

/// Read a string literal starting at `source[start]`, where `source[start]`
/// is the opening quote character (`"`, `'`, or backtick).
pub fn read_string(source: &[char], start: usize, quote: char) -> StringReadResult {
    if source[start] != quote {
        return StringReadResult {
            value: String::new(),
            raw: String::new(),
            end: start,
            has_newline: false,
            is_template: quote == '`',
            expressions: Vec::new(),
            error: Some(format!("Expected {}", quote)),
        };
    }

    let mut i = start + 1;
    let mut value = String::new();
    let mut has_newline = false;
    let mut expressions = Vec::new();
    let is_template = quote == '`';

    while i < source.len() {
        let ch = source[i];

        // Closing quote
        if ch == quote {
            // For template literals, check it's not escaped
            if !is_template || (i > 0 && source[i - 1] != '\\') {
                let raw: String = source[start..=i].iter().collect();
                return StringReadResult {
                    value,
                    raw,
                    end: i + 1,
                    has_newline,
                    is_template,
                    expressions,
                    error: None,
                };
            }
        }

        // Template literal expression ${...}
        if is_template && ch == '$' && i + 1 < source.len() && source[i + 1] == '{' {
            let expr_start = i + 2;
            let (text, end) = read_template_expression(source, expr_start);
            if end == usize::MAX {
                let raw: String = source[start..i].iter().collect();
                return StringReadResult {
                    value,
                    raw,
                    end: source.len(),
                    has_newline,
                    is_template,
                    expressions,
                    error: Some("Unclosed template expression".into()),
                };
            }
            expressions.push(TemplateExpression {
                start: expr_start,
                end,
                expression: text,
            });
            value.push('\x00');
            i = end + 1;
            continue;
        }

        // Escape sequences
        if ch == '\\' {
            let (escaped_val, end, err) = read_escape(source, i);
            if let Some(e) = err {
                let raw: String = source[start..i].iter().collect();
                return StringReadResult {
                    value,
                    raw,
                    end: i,
                    has_newline,
                    is_template,
                    expressions,
                    error: Some(e),
                };
            }
            value.push_str(&escaped_val);
            if is_line_terminator_char(escaped_val.chars().next().unwrap_or('\0')) {
                has_newline = true;
            }
            i = end;
            continue;
        }

        // Line terminators
        if is_line_terminator(ch) {
            if !is_template {
                let raw: String = source[start..i].iter().collect();
                return StringReadResult {
                    value,
                    raw,
                    end: i,
                    has_newline: true,
                    is_template,
                    expressions,
                    error: Some("Unterminated string literal".into()),
                };
            }
            has_newline = true;
        }

        value.push(ch);
        i += 1;
    }

    let raw: String = source[start..].iter().collect();
    StringReadResult {
        value,
        raw,
        end: source.len(),
        has_newline,
        is_template,
        expressions,
        error: Some("Unterminated string literal".into()),
    }
}

fn read_template_expression(source: &[char], start: usize) -> (String, usize) {
    let mut i = start;
    let mut depth = 1;
    let mut text = String::new();

    while i < source.len() {
        let ch = source[i];

        if ch == '{' {
            depth += 1;
        } else if ch == '}' {
            depth -= 1;
            if depth == 0 {
                return (text, i);
            }
        }

        // Skip nested strings
        if ch == '"' || ch == '\'' || ch == '`' {
            let str_result = read_string(source, i, ch);
            let raw_chars: String = str_result.raw.chars().collect();
            text.push_str(&raw_chars);
            i = str_result.end;
            continue;
        }

        text.push(ch);
        i += 1;
    }

    (text, usize::MAX)
}

fn read_escape(source: &[char], start: usize) -> (String, usize, Option<String>) {
    // start is at '\', next char determines escape type
    if start + 1 >= source.len() {
        return ("\\".into(), start + 1, Some("Unterminated escape sequence".into()));
    }

    let ch = source[start + 1];

    // Simple escapes
    let simple = match ch {
        'n' => Some('\n'),
        'r' => Some('\r'),
        't' => Some('\t'),
        'b' => Some('\u{0008}'),
        'f' => Some('\u{000C}'),
        'v' => Some('\u{000B}'),
        '0' => Some('\0'),
        '\\' => Some('\\'),
        '\'' => Some('\''),
        '"' => Some('"'),
        '`' => Some('`'),
        '$' => Some('$'),
        '/' => Some('/'),
        _ => None,
    };

    if let Some(c) = simple {
        return (c.to_string(), start + 2, None);
    }

    // Unicode escape \uXXXX or \u{XXXXX}
    if ch == 'u' {
        if start + 2 < source.len() && source[start + 2] == '{' {
            // \u{XXXXX} — code point escape
            let mut end = start + 3;
            while end < source.len() && source[end] != '}' {
                end += 1;
            }
            if end >= source.len() {
                return (String::new(), start + 2, Some("Unterminated Unicode code point escape".into()));
            }
            let hex: String = source[start + 3..end].iter().collect();
            match u32::from_str_radix(&hex, 16) {
                Ok(cp) if cp <= 0x10FFFF => {
                    if let Some(c) = char::from_u32(cp) {
                        return (c.to_string(), end + 1, None);
                    }
                }
                _ => {}
            }
            return (String::new(), end, Some("Invalid Unicode code point".into()));
        }

        // \uXXXX — fixed-length
        if start + 6 <= source.len() {
            let hex: String = source[start + 2..start + 6].iter().collect();
            if hex.chars().all(|c| c.is_ascii_hexdigit()) {
                if let Ok(cp) = u32::from_str_radix(&hex, 16) {
                    if let Some(c) = char::from_u32(cp) {
                        return (c.to_string(), start + 6, None);
                    }
                }
            }
        }
        return (String::new(), start + 2, Some("Invalid Unicode escape sequence".into()));
    }

    // Hex escape \xXX
    if ch == 'x' {
        if start + 4 <= source.len() {
            let hex: String = source[start + 2..start + 4].iter().collect();
            if hex.chars().all(|c| c.is_ascii_hexdigit()) {
                if let Ok(cp) = u32::from_str_radix(&hex, 16) {
                    if let Some(c) = char::from_u32(cp) {
                        return (c.to_string(), start + 4, None);
                    }
                }
            }
        }
        return (String::new(), start + 2, Some("Invalid hex escape sequence".into()));
    }

    // Octal escape
    if ch.is_ascii_digit() && ch >= '0' && ch <= '7' {
        let mut octal = String::new();
        octal.push(ch);
        let mut j = start + 2;
        while j < source.len() && source[j] >= '0' && source[j] <= '7' && octal.len() < 3 {
            octal.push(source[j]);
            j += 1;
        }
        if let Ok(cp) = u32::from_str_radix(&octal, 8) {
            if cp <= 255 {
                if let Some(c) = char::from_u32(cp) {
                    return (c.to_string(), j, None);
                }
            }
        }
        return (String::new(), j, Some("Octal escape out of range".into()));
    }

    // Line continuation (backslash + line terminator)
    if is_line_terminator(ch) {
        return (String::new(), start + 2, None);
    }

    // Unknown escape — keep character as-is
    (ch.to_string(), start + 2, None)
}

fn is_line_terminator(ch: char) -> bool {
    ch == '\n' || ch == '\r' || ch == '\u{2028}' || ch == '\u{2029}'
}

fn is_line_terminator_char(ch: char) -> bool {
    is_line_terminator(ch)
}

pub fn is_string_delimiter(ch: char) -> bool {
    ch == '"' || ch == '\'' || ch == '`'
}

/// Read a numeric literal starting at source[start].
pub fn read_number(source: &[char], start: usize) -> NumberReadResult {
    let mut i = start;
    let mut value = String::new();
    let mut is_float = false;
    let mut is_big_int = false;
    let mut radix: u32 = 10;

    // Check for hex, octal, binary
    if source[i] == '0' && i + 1 < source.len() {
        let next = source[i + 1].to_ascii_lowercase();
        match next {
            'x' => {
                radix = 16;
                i += 2;
                while i < source.len() && (source[i].is_ascii_hexdigit() || source[i] == '_') {
                    value.push(source[i]);
                    i += 1;
                }
                let raw: String = source[start..i].iter().collect();
                return NumberReadResult { value, raw, end: i, is_float: false, is_big_int, radix, error: None };
            }
            'o' => {
                radix = 8;
                i += 2;
                while i < source.len() && (source[i] >= '0' && source[i] <= '7' || source[i] == '_') {
                    value.push(source[i]);
                    i += 1;
                }
                let raw: String = source[start..i].iter().collect();
                return NumberReadResult { value, raw, end: i, is_float: false, is_big_int, radix, error: None };
            }
            'b' => {
                radix = 2;
                i += 2;
                while i < source.len() && (source[i] == '0' || source[i] == '1' || source[i] == '_') {
                    value.push(source[i]);
                    i += 1;
                }
                let raw: String = source[start..i].iter().collect();
                return NumberReadResult { value, raw, end: i, is_float: false, is_big_int, radix, error: None };
            }
            _ => {}
        }
    }

    // Decimal integer part
    while i < source.len() && (source[i].is_ascii_digit() || source[i] == '_') {
        value.push(source[i]);
        i += 1;
    }

    // Fractional part
    if i < source.len() && source[i] == '.' && i + 1 < source.len() && source[i + 1].is_ascii_digit() {
        is_float = true;
        value.push('.');
        i += 1;
        while i < source.len() && (source[i].is_ascii_digit() || source[i] == '_') {
            value.push(source[i]);
            i += 1;
        }
    }

    // Exponent
    if i < source.len() && (source[i] == 'e' || source[i] == 'E') {
        is_float = true;
        value.push(source[i]);
        i += 1;
        if i < source.len() && (source[i] == '+' || source[i] == '-') {
            value.push(source[i]);
            i += 1;
        }
        while i < source.len() && (source[i].is_ascii_digit() || source[i] == '_') {
            value.push(source[i]);
            i += 1;
        }
    }

    // BigInt suffix
    if i < source.len() && source[i] == 'n' {
        is_big_int = true;
        i += 1;
    }

    let raw: String = source[start..i].iter().collect();
    NumberReadResult { value, raw, end: i, is_float, is_big_int, radix, error: None }
}

/// Read a regex literal /pattern/flags starting at source[start].
pub fn read_regex(source: &[char], start: usize) -> RegexReadResult {
    if source[start] != '/' {
        return RegexReadResult {
            pattern: String::new(),
            flags: String::new(),
            raw: String::new(),
            end: start,
            error: Some("Expected /".into()),
        };
    }

    let mut i = start + 1;
    let mut pattern = String::new();
    let mut in_class = false;

    while i < source.len() {
        let ch = source[i];

        if ch == '\\' {
            pattern.push(ch);
            if i + 1 < source.len() {
                pattern.push(source[i + 1]);
                i += 2;
                continue;
            }
        }

        if ch == '[' { in_class = true; }
        if ch == ']' { in_class = false; }

        if ch == '/' && !in_class {
            i += 1;
            let mut flags = String::new();
            while i < source.len() && "gimsuyd".contains(source[i]) {
                flags.push(source[i]);
                i += 1;
            }
            let raw: String = source[start..i].iter().collect();
            return RegexReadResult { pattern, flags, raw, end: i, error: None };
        }

        if ch == '\n' || ch == '\r' {
            let raw: String = source[start..i].iter().collect();
            return RegexReadResult { pattern, flags: String::new(), raw, end: i, error: Some("Unterminated regex literal".into()) };
        }

        pattern.push(ch);
        i += 1;
    }

    let raw: String = source[start..].iter().collect();
    RegexReadResult { pattern, flags: String::new(), raw, end: source.len(), error: Some("Unterminated regex literal".into()) }
}

/// Escape a string for output, adding surrounding quotes.
pub fn escape_string(s: &str, quote: char) -> String {
    let mut result = String::new();
    result.push(quote);

    for ch in s.chars() {
        match ch {
            '\\' => { result.push_str("\\\\"); }
            c if c == quote => { result.push('\\'); result.push(c); }
            '\n' => { result.push_str("\\n"); }
            '\r' => { result.push_str("\\r"); }
            '\t' => { result.push_str("\\t"); }
            '\u{0008}' => { result.push_str("\\b"); }
            '\u{000C}' => { result.push_str("\\f"); }
            '\u{000B}' => { result.push_str("\\v"); }
            '\0' => { result.push_str("\\0"); }
            c if (c as u32) < 0x20 => {
                result.push_str(&format!("\\x{:02x}", c as u32));
            }
            c if (c as u32) > 0xFFFF => {
                result.push_str(&format!("\\u{{{:x}}}", c as u32));
            }
            c if (c as u32) > 0x7E => {
                result.push_str(&format!("\\u{:04x}", c as u32));
            }
            c => { result.push(c); }
        }
    }

    result.push(quote);
    result
}
